go/login/server/server.go

```diff
@@ -0,0 +1,299 @@
+	ErrInvalidEph = fmt.Errorf("400 bad request: invalid ephemeral Key")
+	// ErrIncorrectTag denotes that received tag is incorrect
+	ErrIncorrectTag = fmt.Errorf("invalid Tag")
+)
+
+// Authorizer responds to an authorization request. The method
+// Allow returns whether an user is allowed to run an action in a host.
+//
+// Some considerations need to be held while implementing this interface:
+// - Allow should accept an empty string as command, denoting that
+// an empty action is passed in the request.
+// - If no user can be obtained from request metadata, an empty string is to be
+// passed as default value.
+// - The host can contain information about both the host type and the host id parsed
+// as <host-type>:<host-id>. the host-type information is optional. if host-type is not
+// provided the string is parsed as <host-id> (that is, we also drop the colon).
+type Authorizer interface {
+	GrantLogin(user string, hostID string, hostIDType string, action string) bool
+}
+
+// AuthorizerFunc type is an adapter to allow the use of ordinary functions as Authorizers.
+type AuthorizerFunc func(user string, hostID string, hostIDType string, action string) bool
+
+// GrantLogin calls a(user, action, host)
+func (a AuthorizerFunc) GrantLogin(user string, hostID string, hostIDType string, action string) bool {
+	return a(user, hostID, hostIDType, action)
+}
+
+// LoginServer is a framework for programming glome-login servers, that will be
+// useful for most use-cases. Creation of the object is managed via the NewLoginServer
+// method. To create one it is necessary to provide an authorizer. Optionally, a series
+// of options can be added at creation.
+type LoginServer struct {
+	// Keys manages the keys used by the server.
+	Keys *KeyManager
+
+	// Unexported fields.
+	auth     Authorizer
+	authLock sync.RWMutex
+	router   *mux.Router
+
+	responseLen uint8
+	userHeader  string
+}
+
+// ServeHTTP implements http.Handler interface.
+func (s *LoginServer) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	s.router.ServeHTTP(w, r)
+}
+
+// Authorizer sets Authorizer function for LoginServer in a concurrentsafe way.
+func (s *LoginServer) Authorizer(a Authorizer) {
+	s.authLock.Lock()
+	s.auth = a
+	s.authLock.Unlock()
+}
+
+// NewServer creates a new server with provided Authorizer and, optionally, selected options
+func NewLoginServer(a Authorizer, options ...func(*LoginServer) error) (*LoginServer, error) {
+	srv := LoginServer{
+		auth:        a,
+		Keys:        NewKeyManager(),
+		responseLen: MaxResponseSize,
+	}
+	srv.router = srv.newRouter()
+
+	for _, option := range options {
+		if err := option(&srv); err != nil {
+			return nil, err
+		}
+	}
+
+	return &srv, nil
+}
+
+// ResponseLen is an option to be provided to NewServer on creation. Its sets response len to
+// n characters long response.
+func ResponseLen(n uint8) func(srv *LoginServer) error {
+	return func(srv *LoginServer) error {
+		if 1 > n || MaxResponseSize < n {
```

Comment:
Did you mean `n == 0 || ...`?

Reply:
imho opinion 0 < n conveys more information on reading. Still, change the line to 
```
if !(0 < n && n < MaxResponseSize) {                                                                            
```
So it is more readable.